src/engine/engine.cc

```diff
@@ -67,4 +67,26 @@ Engine* Engine::Get() {
   static Engine *inst = _GetSharedRef().get();
   return inst;
 }
+
+void Engine::PushAsyncPtr(AsyncFnPtr exec_fn_ptr, const std::shared_ptr<void>& param,
+                          Context exec_ctx, std::vector<VarHandle> const& const_vars,
+                          std::vector<VarHandle> const& mutable_vars,
+                          FnProperty prop, int priority,
+                          const char* opr_name, bool wait) {
+  auto exec_fn = [exec_fn_ptr, param](RunContext rctx,
```

Comment:
can we do  a by ref capture here.

Reply:
I think we cannot do a ref capture here. This is where we want to do a copy of the shared_ptr which increases its ref count. When the exec_fn_ptr is called in another thread, we need to make sure param is still valid